replica/table.cc

```diff
@@ -1300,130 +1272,118 @@ seal_snapshot(sstring jsondir) {
+        ex = std::current_exception();
+    }
+    co_await out.close();
+
+    if (ex) {
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
+
+    co_await io_check(sync_directory, std::move(jsondir));
 }
 
 future<> table::write_schema_as_cql(database& db, sstring dir) const {
     std::ostringstream ss;
+
+    this->schema()->describe(db, ss);
+
+    auto schema_description = ss.str();
+    auto schema_file_name = dir + "/schema.cql";
+    auto f = co_await open_checked_file_dma(general_disk_error_handler, schema_file_name, open_flags::wo | open_flags::create | open_flags::truncate);
+    auto out = co_await make_file_output_stream(std::move(f));
+    std::exception_ptr ex;
+
     try {
-        this->schema()->describe(db, ss);
+        co_await out.write(schema_description.c_str(), schema_description.size());
+        co_await out.flush();
     } catch (...) {
-        return make_exception_future<>(std::current_exception());
+        ex = std::current_exception();
     }
-    auto schema_description = ss.str();
-    auto schema_file_name = dir + "/schema.cql";
-    return open_checked_file_dma(general_disk_error_handler, schema_file_name, open_flags::wo | open_flags::create | open_flags::truncate).then([schema_description = std::move(schema_description)](file f) {
-        return make_file_output_stream(std::move(f)).then([schema_description  = std::move(schema_description)] (output_stream<char>&& out) mutable {
-            return do_with(std::move(out), [schema_description  = std::move(schema_description)] (output_stream<char>& out) {
-                return out.write(schema_description.c_str(), schema_description.size()).then([&out] {
-                   return out.flush();
-                }).then([&out] {
-                   return out.close();
-                });
-            });
+    co_await out.close();
+
+    if (ex) {
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
+}
+
+// Runs the orchestration code on an arbitrary shard to balance the load.
+future<> table::snapshot_on_all_shards(sharded<database>& sharded_db, const std::vector<foreign_ptr<lw_shared_ptr<table>>>& table_shards, sstring name) {
+    auto jsondir = table_shards[this_shard_id()]->_config.datadir + "/snapshots/" + name;
+    auto orchestrator = std::hash<sstring>()(jsondir) % smp::count;
+
+    co_await smp::submit_to(orchestrator, [&] () -> future<> {
+        auto& t = *table_shards[this_shard_id()];
+        auto s = t.schema();
+        tlogger.debug("Taking snapshot of {}.{}: directory={}", s->ks_name(), s->cf_name(), jsondir);
+
+        std::vector<table::snapshot_file_set> file_sets;
+        file_sets.reserve(smp::count);
+
+        co_await coroutine::parallel_for_each(boost::irange(0u, smp::count), [&] (unsigned shard) -> future<> {
+            file_sets.emplace_back(co_await smp::submit_to(shard, [&] {
+                return table_shards[this_shard_id()]->take_snapshot(sharded_db.local(), jsondir);
```

Comment:
I think vector's emplace back is not thread safe.

Better to place snapshot_file_set in the idx of the originating shard.

Reply:
please disregard my comment. we're doing the insertion via the coordinator.